### PR TITLE
fix(@aws-amplify/pushnotification): Upgrade Firebase messaging Android package

### PR DESCRIPTION
**Pull request by Samaritan1011001**
#### Description of changes

This PR upgrades the firebase package version used in the Android part of the pushnotification category. The latest firebase version involves breaking changes for us and such instances are updated as well. They are, [`getToken`](https://firebase.google.com/docs/reference/android/com/google/firebase/iid/FirebaseInstanceId#getInstanceId()) and [`onNewToken` ](https://firebase.google.com/docs/reference/android/com/google/firebase/iid/FirebaseInstanceIdService#public-void-ontokenrefresh)APIs. 

This is **not a breaking change** to customers. It only forces us to update some **internal** functions in the pushnotificaiton category as shown in the PR.

In addition, the `RNPushNotificationDeviceIDService`  service has been removed since the new Firebase API requires us to overrride the `onNewToken` method from `FirebaseMessagingService` as seen in `RNPushNotificationMessagingService`. 

#### Issue #, if available

https://github.com/aws-amplify/amplify-js/issues/8506
https://github.com/aws-amplify/amplify-js/issues/8950

#### Description of how you validated changes
1. Manual testing
2. Ran an integration test using the sample app found here- https://github.com/aws-amplify/amplify-js-samples-staging/tree/master/samples/react-native/push-notifications/PushNotificationApp
 A separate PR for these tests will be submitted separately

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [ ] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/pushnotification/android/src/main/java/com/amazonaws/amplify/pushnotification/RNPushNotificationModule.java [hunk 1]
```diff
@@ -19,6 +19,8 @@
 import android.content.IntentFilter;
 import android.content.BroadcastReceiver;
 
+import androidx.annotation.NonNull;
```
Comment (elorzafe) [inline packages/pushnotification/android/src/main/java/com/amazonaws/amplify/pushnotification/RNPushNotificationModule.java:22]: why is this needed?

File: packages/pushnotification/android/src/main/java/com/amazonaws/amplify/pushnotification/RNPushNotificationMessagingService.java [hunk 2]
```diff
@@ -167,4 +168,54 @@ private Bundle convertDataToBundle(RemoteMessage message) {
...
 
+	/**
+	 * Called if InstanceID token is updated. This may occur if the security of
+	 * the previous token had been compromised. Note that this is called when the InstanceID token
+	 * is initially generated so this is where you would retrieve the token.
+	 */
+	@Override
+	public void onNewToken(String refreshedToken) {
+		// Get updated InstanceID token.
+		Log.v(LOG_TAG, "Refreshed token: " + refreshedToken);
```
Comment (elorzafe) [inline packages/pushnotification/android/src/main/java/com/amazonaws/amplify/pushnotification/RNPushNotificationMessagingService.java:179]: Should this be for debug?

**Review by elorzafe — COMMENTED** [review]
A couple of questions

Overall looks good, just want to confirm.

This will require a docs update?

File: packages/pushnotification/android/src/main/java/com/amazonaws/amplify/pushnotification/RNPushNotificationModule.java [hunk 3]
```diff
@@ -64,9 +68,15 @@ public void initialize() {
     }
 
     @ReactMethod
-    public void getToken(Callback callback) {
-        String token =  FirebaseInstanceId.getInstance().getToken();
-        Log.i(LOG_TAG, "getting token" + token);
-        callback.invoke(token);
+    public void getToken(final Callback callback) {
+        final Task<String> taskToken =  FirebaseMessaging.getInstance().getToken();
+			taskToken.addOnSuccessListener(new OnSuccessListener<String>() {
```
Comment (cshfang) [inline packages/pushnotification/android/src/main/java/com/amazonaws/amplify/pushnotification/RNPushNotificationModule.java:73]: What happens if getToken fails?

File: packages/pushnotification/android/src/main/java/com/amazonaws/amplify/pushnotification/RNPushNotificationMessagingService.java (hunk 2 above)
Comment (Samaritan1011001) [reply packages/pushnotification/android/src/main/java/com/amazonaws/amplify/pushnotification/RNPushNotificationMessagingService.java:179]: Yeah, it should be `Log.i` like the rest of the places in this file. Changed ✔️

File: packages/pushnotification/android/src/main/java/com/amazonaws/amplify/pushnotification/RNPushNotificationModule.java (hunk 1 above)
Comment (Samaritan1011001) [reply packages/pushnotification/android/src/main/java/com/amazonaws/amplify/pushnotification/RNPushNotificationModule.java:22]: It's not. Bad import. Removed ✔️

File: packages/pushnotification/android/src/main/java/com/amazonaws/amplify/pushnotification/RNPushNotificationModule.java (hunk 3 above)
Comment (Samaritan1011001) [reply packages/pushnotification/android/src/main/java/com/amazonaws/amplify/pushnotification/RNPushNotificationModule.java:73]: With this current state. It silently fails. The previous code didn't handle it as well so I left as is. But I'll add it now. Thank you for catching that!

File: packages/pushnotification/android/src/main/java/com/amazonaws/amplify/pushnotification/RNPushNotificationModule.java (hunk 1 above)
Comment (Samaritan1011001) [reply packages/pushnotification/android/src/main/java/com/amazonaws/amplify/pushnotification/RNPushNotificationModule.java:22]: This is needed for the `onSuccess` and `onError` does not return Null values. I added proper usage in the latest commit. Thanks for catching this.

**Comment by sammartinez** [comment]
For semantic wise, if we are calling this out as a breaking change, we will need to adjust this from being a `fix` to `fix!` and bump the version of the pushnotifications package

File: packages/pushnotification/android/src/main/java/com/amazonaws/amplify/pushnotification/RNPushNotificationMessagingService.java [hunk 4]
```diff
@@ -167,4 +168,54 @@ private Bundle convertDataToBundle(RemoteMessage message) {
...
+	 * the previous token had been compromised. Note that this is called when the InstanceID token
+	 * is initially generated so this is where you would retrieve the token.
+	 */
+	@Override
+	public void onNewToken(String refreshedToken) {
+		// Get updated InstanceID token.
+		Log.i(LOG_TAG, "Refreshed token: " + refreshedToken);
+
+		final Bundle bundle = createBundleWithToken(refreshedToken);
+		// We need to run this on the main thread, as the React code assumes that is true.
```
Comment (sammartinez) [inline packages/pushnotification/android/src/main/java/com/amazonaws/amplify/pushnotification/RNPushNotificationMessagingService.java:182]: This is specific to React Native correct? 

```suggestion
		// We need to run this on the main thread, as the React Native code assumes that is true.
```

File: packages/pushnotification/android/src/main/java/com/amazonaws/amplify/pushnotification/RNPushNotificationMessagingService.java [hunk 5]
```diff
@@ -167,4 +168,54 @@ private Bundle convertDataToBundle(RemoteMessage message) {
...
+		// Namely, DevServerHelper constructs a Handler() without a Looper, which triggers:
+		// "Can't create handler inside thread that has not called Looper.prepare()"
+		Handler handler = new Handler(Looper.getMainLooper());
+		handler.post(new Runnable() {
+			public void run() {
+				// Construct and load our normal React JS code bundle
```
Comment (sammartinez) [inline packages/pushnotification/android/src/main/java/com/amazonaws/amplify/pushnotification/RNPushNotificationMessagingService.java:188]: Do we need JS there? 

```suggestion
				// Construct and load our normal React code bundle
```

**Review by sammartinez — COMMENTED** [review]
Couple callouts as well

File: packages/pushnotification/android/src/main/java/com/amazonaws/amplify/pushnotification/RNPushNotificationModule.java [hunk 6]
```diff
@@ -64,9 +67,22 @@ public void initialize() {
...
-        Log.i(LOG_TAG, "getting token" + token);
-        callback.invoke(token);
+    public void getToken(final Callback callback) {
+        final Task<String> taskToken =  FirebaseMessaging.getInstance().getToken();
+            taskToken.addOnCompleteListener(new OnCompleteListener<String>() {
+                @Override
+                public void onComplete(@NonNull Task<String> task) {
+                    if (task.isSuccessful()) {
+                        String token = task.getResult();
+                        Log.i(LOG_TAG, "getting token " + token);
```
Comment (cshfang) [inline packages/pushnotification/android/src/main/java/com/amazonaws/amplify/pushnotification/RNPushNotificationModule.java:77]: Nit: If successful, maybe "got token" is more accurate here?
```suggestion
                        Log.i(LOG_TAG, "got token " + token);
```

File: packages/pushnotification/android/src/main/java/com/amazonaws/amplify/pushnotification/RNPushNotificationModule.java [hunk 7]
```diff
@@ -64,9 +67,22 @@ public void initialize() {
...
+					    callback.invoke(token);
```
Comment (cshfang) [inline packages/pushnotification/android/src/main/java/com/amazonaws/amplify/pushnotification/RNPushNotificationModule.java:78]: Nit: Looks like there's an indentation issue here

File: packages/pushnotification/android/src/main/java/com/amazonaws/amplify/pushnotification/RNPushNotificationModule.java [hunk 8]
```diff
@@ -64,9 +67,22 @@ public void initialize() {
...
+                    } else {
+                        Exception exception = task.getException();
+                        Log.e(LOG_TAG, "Error getting token:" + exception);
+
```
Comment (cshfang) [inline packages/pushnotification/android/src/main/java/com/amazonaws/amplify/pushnotification/RNPushNotificationModule.java:82]: Nit: Stray empty line

File: packages/pushnotification/android/src/main/java/com/amazonaws/amplify/pushnotification/RNPushNotificationModule.java [hunk 9]
```diff
@@ -64,9 +67,22 @@ public void initialize() {
...
+            taskToken.addOnCompleteListener(new OnCompleteListener<String>() {
+                @Override
+                public void onComplete(@NonNull Task<String> task) {
+                    if (task.isSuccessful()) {
+                        String token = task.getResult();
+                        Log.i(LOG_TAG, "getting token " + token);
+					    callback.invoke(token);
+                    } else {
+                        Exception exception = task.getException();
+                        Log.e(LOG_TAG, "Error getting token:" + exception);
```
Comment (cshfang) [inline packages/pushnotification/android/src/main/java/com/amazonaws/amplify/pushnotification/RNPushNotificationModule.java:81]: Nit: May want an extra space after the `:`
```suggestion
                        Log.e(LOG_TAG, "Error getting token: " + exception);
```

**Review by cshfang — COMMENTED** [review]
Lgtm overall. Got some nit comments mostly around formatting.

File: packages/pushnotification/android/src/main/java/com/amazonaws/amplify/pushnotification/RNPushNotificationMessagingService.java (hunk 4 above)
Comment (Samaritan1011001) [reply packages/pushnotification/android/src/main/java/com/amazonaws/amplify/pushnotification/RNPushNotificationMessagingService.java:182]: That's right. This function has been moved from the other class that was used by the previous firebase version called `RNPushNotificationDeviceIDService`. 
And yeah you are right. It is React Native specific. Thanks!

**Comment by Samaritan1011001** [comment]
This is not a breaking change to customers. It only forces us to update some internal functions in the pushnotificaiton category as shown in the PR.

**Review by calebpollman — COMMENTED** [review]
Thanks for doing this @Samaritan1011001!

Have some questions

File: packages/pushnotification/android/src/main/java/com/amazonaws/amplify/pushnotification/RNPushNotificationMessagingService.java [hunk 10]
```diff
@@ -167,4 +168,54 @@ private Bundle convertDataToBundle(RemoteMessage message) {
...
+				ReactContext context = mReactInstanceManager.getCurrentReactContext();
+				// If it's constructed, send a notification
+				if (context != null) {
+					sendRegistrationToken((ReactApplicationContext) context, bundle);
+				} else {
+					// Otherwise wait for construction, then send the notification
+					mReactInstanceManager.addReactInstanceEventListener(new ReactInstanceManager.ReactInstanceEventListener() {
+						public void onReactContextInitialized(ReactContext context) {
+							sendRegistrationToken((ReactApplicationContext) context, bundle);
+						}
```
[the remark this answers is not in the document]
Comment (Samaritan1011001) [reply packages/pushnotification/android/src/main/java/com/amazonaws/amplify/pushnotification/RNPushNotificationMessagingService.java:199]: > Concerned that it could be registered multiple times

By registered do you mean `ReactInstanceEventListener`  getting registered multiple times? 
This code has been moved from the previous [location](https://github.com/aws-amplify/amplify-js/blob/959d908583c2ee4e0cc033efe254ab24cbe83692/packages/pushnotification/android/src/main/java/com/amazonaws/amplify/pushnotification/RNPushNotificationDeviceIDService.java) and remains untouched. 
I believe this only gets called and registered when the `context` is null indicating the app has been killed.

[the remark this answers is not in the document]
Comment (calebpollman) [reply packages/pushnotification/android/src/main/java/com/amazonaws/amplify/pushnotification/RNPushNotificationMessagingService.java:199]: Mean that `onReactContextInitialized` could be registered multiple times as `onNewToken` could be called more than once while the `context` is `null`

File: packages/pushnotification/android/src/main/java/com/amazonaws/amplify/pushnotification/RNPushNotificationModule.java [hunk 11]
```diff
@@ -64,9 +67,22 @@ public void initialize() {
...
+    public void getToken(final Callback callback) {
+        final Task<String> taskToken =  FirebaseMessaging.getInstance().getToken();
+            taskToken.addOnCompleteListener(new OnCompleteListener<String>() {
+                @Override
+                public void onComplete(@NonNull Task<String> task) {
+                    if (task.isSuccessful()) {
+                        String token = task.getResult();
+                        Log.i(LOG_TAG, "getting token " + token);
+					    callback.invoke(token);
+                    } else {
```
[the remark this answers is not in the document]
Comment (Samaritan1011001) [reply packages/pushnotification/android/src/main/java/com/amazonaws/amplify/pushnotification/RNPushNotificationModule.java:79]: This function is only ever used in one place internally and not exposed to customers. Even if I add an error callback here, I can only log the error in `[Pushnotification.ts](https://github.com/aws-amplify/amplify-js/blob/959d908583c2ee4e0cc033efe254ab24cbe83692/packages/pushnotification/src/PushNotification.ts#L142-L147)`. If you think that's more desirable i can add that.

[the remark this answers is not in the document]
Comment (calebpollman) [reply packages/pushnotification/android/src/main/java/com/amazonaws/amplify/pushnotification/RNPushNotificationModule.java:79]: Why can we only log the error?

[the remark this answers is not in the document]
Comment (Samaritan1011001) [reply packages/pushnotification/android/src/main/java/com/amazonaws/amplify/pushnotification/RNPushNotificationModule.java:79]: This function is not exposed to customers. It is to only get the token once when initialized.

File: packages/pushnotification/android/src/main/java/com/amazonaws/amplify/pushnotification/RNPushNotificationMessagingService.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (Samaritan1011001) [reply packages/pushnotification/android/src/main/java/com/amazonaws/amplify/pushnotification/RNPushNotificationMessagingService.java]: Great catch and thanks for the help!

File: packages/pushnotification/android/src/main/java/com/amazonaws/amplify/pushnotification/RNPushNotificationModule.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (Samaritan1011001) [reply packages/pushnotification/android/src/main/java/com/amazonaws/amplify/pushnotification/RNPushNotificationModule.java:78]: Added an error callback so it can be logged on the JS side

File: packages/pushnotification/android/src/main/java/com/amazonaws/amplify/pushnotification/RNPushNotificationMessagingService.java [hunk 12]
```diff
@@ -167,4 +168,55 @@ private Bundle convertDataToBundle(RemoteMessage message) {
...
+				ReactContext context = mReactInstanceManager.getCurrentReactContext();
+				// If it's constructed, send a notification
+				if (context != null) {
+					sendRegistrationToken((ReactApplicationContext) context, bundle);
+				} else {
+					// Otherwise wait for construction, then send the notification
+					mReactInstanceManager.addReactInstanceEventListener(new ReactInstanceManager.ReactInstanceEventListener() {
+						public void onReactContextInitialized(ReactContext context) {
+							sendRegistrationToken((ReactApplicationContext) context, bundle);
+                            mReactInstanceManager.removeReactInstanceEventListener(this);
```
Comment (calebpollman) [inline packages/pushnotification/android/src/main/java/com/amazonaws/amplify/pushnotification/RNPushNotificationMessagingService.java:199]: Nit: indentation looks off

**Review by calebpollman — APPROVED** [review]
LGTM :ship:

**Review by cshfang — APPROVED** [review]
🚢

File: packages/pushnotification/android/src/main/java/com/amazonaws/amplify/pushnotification/RNPushNotificationMessagingService.java (hunk 12 above)
Comment (Samaritan1011001) [reply packages/pushnotification/android/src/main/java/com/amazonaws/amplify/pushnotification/RNPushNotificationMessagingService.java:199]: vscode says there isn't anything wrong 😅

**Review by elorzafe — APPROVED** [review]
LGTM thanks @Samaritan1011001 🌮
